### PR TITLE
AWS Interactive Video Service (IVS) Demo, and Personalised Discount addition to the retail demo

### PR DESCRIPTION
**Pull request by AkramDweikat**
… Summary as follow:

* IVS: Addition of IVS container, IVS cloud formation.
* UI: Adding UI Tab and page for the streams, its products, recommendations and discounts.
* UI: Adding Reset Amazon Personalize button on the user profile.
* UI: Refactoring for the menu.
* Amazon Personalize: Addition of new campaign for discounts, update to the container service code.
* Amazon Personalize: Additional Jupyter notebook, remove dependancy on bucket to stage data,  instead generate it from script.
* Update to the readmes and documentations.
This work was done by Dae.mn Team.
Co-authored-by: Akram Dweikat <akram.dweikat@gmail.com>
Co-authored-by: Akram Dweikat  <akram.dweikat@dae.mn>
Co-authored-by: Damien Jade Duff <damien.duff@daemonsolutions.com>
Co-authored-by: daemon-joe <joe.major@dae.mn>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/recommendations/src/recommendations-service/app.py [hunk 1]
```diff
@@ -433,16 +562,118 @@ def experiment_outcome():
...
+        arn = event_tracker['eventTrackerArn']
+        describe_response = personalize.describe_event_tracker(eventTrackerArn=arn)
+        tracking_id = describe_response['eventTracker']['trackingId']
+        if tracking_id == event_tracking_id:
+            return arn
+    return None
+
+
+@app.route('/reset/realtime', methods=['POST'])
+def reset_realtime():
```
Comment (james-jory) [inline src/recommendations/src/recommendations-service/app.py:598]: This seems out of scope for the IVS integration. What is the requirement here?

File: src/videos/src/videos-service/app.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+SUBTITLE_FORMAT = 'srt'
+LOCAL_VIDEO_DIR = '/app/video-files/'
+
+# -- Parameterised ffmpeg commands
+FFMPEG_STREAM_CMD = """ffmpeg -loglevel panic -hide_banner -re -stream_loop -1 -i \"{video_filepath}\" \
+                           -r 30 -c:v copy -f flv rtmps://{ingest_endpoint}:443/app/{stream_key} -map 0:s -f {subtitle_format} -"""
+FFMPEG_SUBS_COMMAND = "ffmpeg -i \"{video_filepath}\" \"{subtitle_path}\""
+
+# -- Default stream details
+# These details will be returned by the API if the 'use default IVS streams' option is chosen for deployment.
```
Comment (james-jory) [inline src/videos/src/videos-service/app.py:36]: Where do the playback URLs come from? Who owns and maintains them? Can they change?

File: src/videos/src/videos-service/app.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        global stream_details
+        stream_details = DEFAULT_STREAM_DETAILS
+
+
+def stream(s3_video_key, ivs_channel_arn, channel_id):
+    """
+        Starts the stream for a given video file and IVS channel. The video file is streamed on a loop using ffmpeg, and
+        any attached metadata (from the subtitles embedded in the video file) is sent to the channel's `put_metadata`
+        endpoint.
+    """
```
Comment (james-jory) [inline src/videos/src/videos-service/app.py:197]: So we are streaming from this Docker container? Doesn't IVS do the streaming? Sorry for the naive question.

**Review by james-jory — CHANGES_REQUESTED** [review]
This is a massive PR and probably should have been done in phases. But we'll get through it. Excited to see IVS integrated! Some great work here.

General questions:

- Why has the delete personalize resources CR been deleted? How will Personalize resources be cleaned up when the user deletes the CFN stack?
- Should the 1.3 personalized discounts notebook be 1.2 and the cleanup notebook renamed to 1.3? The idea of the cleanup notebook is to cleanup after you've finished the other notebook(s) in the folder.
- The discount notebook seems to be more focused on validating the interaction generation methodology than as an educational tool for customers. For example, all of the steps a customer would normally want to learn on how to use the Personalize service is hidden away in the lambda function. I'm also wondering if the 4th campaign is necessary. It's not completely clear what its purpose is. Could contextual metadata have been used instead with the existing campaigns? 

See more detailed questions throughout PR.

Thanks!

**Comment by james-jory** [comment]
Additional questions:

- The IVS integration looks to only support pre-recorded videos. How can a live stream be done using the integration?  
- Will there be a workshop notebook added to walk a developer through how to add IVS to the Retail Demo Store and run a live stream?
- Given the scope of changes, have you done regression testing against the full application?
- The delivery report (separate doc) indicates that CFN deploys sometimes fail and that it is a pre-existing issue. Please provide details on failures you have experienced so they can be fixed. There are no known/tracked issues with deployment at this time.

Thanks!

File: src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py [hunk 4]
```diff
@@ -284,526 +657,499 @@ def create_recent_purchase_filter(dataset_group_arn, ssm_parameter_name):
...
+            }
 
-    return {
-        'statusCode': 200,
-        'body': json.dumps('Related product, Product recommendation, and Personalized reranking campaigns fully provisioned!')
-    }
\ No newline at end of file
+    else:
+        # We have a Cloud Formation Event - the event contains info about whether it wants to update, create or delete
+        cloudformation_helper(event, context)
```
Comment (james-jory) [inline src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py:1155]: Changing this function to a custom resource is going to change the CFN deployment time from 25-30 minutes to 2+ hours (if enabled), right? I have concerns about the customer experience with deploying this project for developers as this length of time for a deployment would be very unusual. 

Also, this is going to break deployments using our (AWS) internal tool for workshops that will timeout deployments that take this long.

**Review by james-jory — CHANGES_REQUESTED** [review]

File: generators/generate_interactions_personalize.py [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+# Percentages of each event type to generate
+product_added_percent = .08
+cart_viewed_percent = .05
+checkout_started_percent = .02
+order_completed_percent = .01
+
+CATEGORY_AFFINITY_PROBS = np.array(CATEGORY_AFFINITY_PROBS)
+
+
+def generate_data():
```
Comment (james-jory) [inline generators/generate_interactions_personalize.py:89]: This implementation adds a third algorithm/approach for creating interactions in the project. 

- The original interaction generator in the Personalize notebook.
- The generator framework introduced under generators.
- This implementation.

The plan is to consolidate and move towards a single algorithm rather than splinter into multiple approaches. Can these improvements be incorporated into the generators algorithm?

File: aws/cloudformation-templates/deployment-support.yaml [hunk 6]
```diff
@@ -159,48 +162,33 @@ Resources:
...
-                  - personalize:DescribeDatasetGroup
-                  - personalize:DescribeDatasetImportJob
-                  - personalize:DescribeSolution
-                  - personalize:DescribeEventTracker
-                  - personalize:CreateCampaign
-                  - personalize:CreateDataset
-                  - personalize:CreateEventTracker
-                  - personalize:CreateFilter
-                Resource: 
-                  - !Sub 'arn:aws:personalize:${AWS::Region}:${AWS::AccountId}:*/retaildemostore*'
```
[the remark this answers is not in the document]
Comment (alexchirayath) [reply aws/cloudformation-templates/deployment-support.yaml:188]: +1 on this comment. Scoping down permissions whenever possible is for security around the application. If you need all permissions or on all resources, please break down into multiple policies which will be scoped accordingly. If there are any * policies, please provide a comment linking to AWS documentation that requires *

File: src/products/src/products-service/data/products.yaml [hunk 7]
```diff
@@ -430,4 +430,262 @@
...
   price: 9.99
-  image: "12.jpg"
\ No newline at end of file
+  image: "12.jpg"
+- id: 1000567890
+  name: Eyeshadow Palette - Set of 3 Palettes
+  category: beauty
+  style: makeup
+  description: Perfect for trialling different shades, each palette contains 4 different shades.
+  price: 145.00
```
Comment (alexchirayath) [inline src/products/src/products-service/data/products.yaml:439]: What is the reasoning behind the item IDs?
I see some really large IDS (1000567891) and some that are close to the existing items id (101,102...)

File: aws/cloudformation-templates/deployment-support.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:188]: No problem, will scope these down.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 8]
```diff
@@ -243,27 +231,37 @@ Resources:
           - "PersonalizePreCreateScheduledRule"
           - "Arn"
 
-  ####################### Personalize Resource Delete Custom Resource #######################
+  # Custom resource to launch IVS create channels function
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:234]: Typo - Will amend

File: aws/cloudformation-templates/services/service/service.yaml [hunk 9]
```diff
@@ -153,14 +171,72 @@ Resources:
...
+                  - ivs:GetChannel
+                  - ivs:GetStream
+                  - ivs:PutMetadata
+                Resource:
+                  - !Sub 'arn:aws:ivs:${AWS::Region}:${AWS::AccountId}:channel/*'
+              - Effect: Allow
+                Action:
+                  - ivs:GetStreamKey
+                Resource:
+                  - !Sub 'arn:aws:ivs:${AWS::Region}:${AWS::AccountId}:stream-key/*'
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/services/service/service.yaml:231]: Yes. Will minimise the scope on all of these.

File: src/products/src/products-service/data/products.yaml [hunk 10]
```diff
@@ -430,4 +430,262 @@
   style: fishing
   description: Every tackle box needs a silver and red reflective fishing lure.
   price: 9.99
-  image: "12.jpg"
\ No newline at end of file
+  image: "12.jpg"
+- id: 1000567890
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply src/products/src/products-service/data/products.yaml:434]: One set of images were obtained with IDs from a separate AWS demo https://github.com/aws-samples/amazon-ivs-ecommerce-web-demo and we carried their IDs over.
The second set was numbered according to the video for which they were acquired. 
However, this can be changed for internal consistency if required.

File: src/recommendations/src/recommendations-service/app.py [hunk 11]
```diff
@@ -286,122 +320,217 @@ def recommendations():
...
+        ranked_items = resolver.get_items(
+            user_id=user_id,
+            product_list=unranked_items
+        )
+
+    response_items = []
+    for ranked_item in ranked_items:
+
+        # Unlike with /recommendations and /related we are not hitting the products API to get product info back
+        # One would hope the caller would have it since the products are coming from there
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply src/recommendations/src/recommendations-service/app.py:449]: This is correct, we will update the comment. 

Note that currently for the discounts we are doing it the way the comment says, but not the search reranking.

File: src/recommendations/src/recommendations-service/app.py [hunk 12]
```diff
@@ -286,122 +320,217 @@ def recommendations():
...
+        item = item_map.get(ranked_item.get('itemId'))
+
+        if 'experiment' in ranked_item:
+
+            # We do not want to override an experiment if info already there
+            # e.g. if client is pulling recommendations then reranking them for example
+            # - this should not happen as the first recommendation pull would already be ranked
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply src/recommendations/src/recommendations-service/app.py:456]: The comment was trying to explain a potential corner case but you're absolutely right that it is unlikely anyone would try that - will delete the comment to avoid confusion.

File: src/recommendations/src/recommendations-service/app.py [hunk 13]
```diff
@@ -286,122 +320,217 @@ def recommendations():
...
-                ranked_items = resolver.get_items(
-                    user_id = user_id, 
-                    product_list = unranked_items
-                )
+                product_url += 'exp=' + ranked_item['experiment']['correlationId']
 
-            response_items = []
-            for ranked_item in ranked_items:
-                item = item_map.get(ranked_item.get('itemId'))
+                item['old_url'] = item['url']
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply src/recommendations/src/recommendations-service/app.py:472]: In the existing reranking code, which we endeavoured to keep as intact as possible considering we also repurposed it for discounts, the item URL is retrieved and the experiment ID added. This is presumably so that downstream it can be seen and picked up (though client-side mostly the experiment dictionary is used). We attempted to remove references to the experiment if the product was not involved in the discount recommendation but clearly the better approach overall would be to leave it in. Another approach would be to introduce into the experimentation a separate resolver type to handle with the "picking" type experiment.

File: src/recommendations/src/recommendations-service/app.py (hunk 1 above)
Comment (daemon-joe) [reply src/recommendations/src/recommendations-service/app.py:598]: Since this demo is going to be running for a number of days at a conference provided functionality to reset the campaigns to their default state in case the output of the campaigns was to drift over the course of a number of demos. 

This reset functionality used to be possible with Amazon Personalize by deleting the so-called "EVENT_INTERACTIONS" dataset but this dataset seems deprecated so we experimented with reset options and ultimately needed to:
- Delete dataset group & all content
- Recreate dataset group with a new name
- Recreate solutions and campaigns with new names

Hopefully this provides some background for some of the other changes (e.g. why we merged delete and create functionality into one lambda).

Following yesterday's (22nd Oct) call we are also aware of the blue-green deployment approach that other teams have been working on - we should discuss this further on our next call so we can unify our approach to this sort of functionality.

File: src/videos/src/videos-service/app.py (hunk 2 above)
Comment (daemon-joe) [reply src/videos/src/videos-service/app.py:36]: We were asked to ensure that streams containing realistic content (ie. the vlog-style videos we have purchased) would be available to all users. Due to licensing issues, we are unable to distribute the video files to other users of the demo so we have been asked to deploy persistent streams which all users can consume using the IVS ingest URL. The AWS Retail team will maintain them going forward. 

The streams can change if there is a full redeployment of the underlying CloudFormation stack. In this case the default streams would need to be updated here. However, stream URLs would not change on new container deployments (we are using the same CD pipeline as is used in this repo).

The AWS retail team is in the process of obtaining a domain for the IVS demo; we could possibly use that to make the default URLs persistent. One for discussion on a call.

File: src/videos/src/videos-service/app.py (hunk 3 above)
Comment (daemon-joe) [reply src/videos/src/videos-service/app.py:197]: IVS provides the channel from which video content can be consumed, but the video must be streamed to the channel from an external source (IVS is the middle-man that allows us to scale to N subscribers along with its other features). This is either using live-streaming software which a user would run as part of a broadcast setup (beyond our scope) or using pre-recorded video. Our implementation is based on the ‘Streaming a Recorded Video with FFmpeg’ section of this guide: https://docs.aws.amazon.com/ivs/latest/userguide/GSIVS-live-streaming.html

File: src/products/src/products-service/data/products.yaml (hunk 7 above)
Comment (daemon-joe) [reply src/products/src/products-service/data/products.yaml:439]: I've replied to a similar question [here](https://github.com/aws-samples/retail-demo-store/pull/96#discussion_r510249803). 

We are happy to standardise for consistency if required

File: generators/generate_interactions_personalize.py (hunk 5 above)
Comment (daemon-joe) [reply generators/generate_interactions_personalize.py:89]: No doubt it is possible. 

The challenge would be the same - to allow for balance across products & categories (if interactions for a particular category are "spread too thin" we can end up with nothing recommended from the category and some items can get forgotten). If more products and categories are in the process of being added to the demo, this may be less of an issue? Note that the old interactions generator also considers gender as a factor which we would need to bring in to the new one (but this was one cause of imbalance). 

Alternatively, we could fold any changes in when the original interaction generator is folded in (it seems this is planned?).

File: src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py (hunk 4 above)
Comment (daemon-joe) [reply src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py:1155]: We agree that 2+ hours would be too long to wait for CloudFormation, though this is not quite what happens. 

As was the case previously, once Personalize starts building, CloudFormation sees it as deployed. Indeed, if we allowed CloudFormation to wait for the Custom Resource Lambda to finish polling, CloudFormation itself would time out. So what we do in the Custom Resource Lambda is return immediately and post the event rule as was the case with the previous solution. CloudFormation finishes its deployment and Personalize continues to train by polling the Lambda with the event rule (if Personalize deployment is enabled). 

That event rule is also used for "reset button" functionality to poll the Lambda when changes are requested to the Personalize deployment.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 14]
```diff
@@ -221,9 +209,9 @@ Resources:
     Type: 'AWS::Events::Rule'
     Properties:
       Name: 'RetailDemoStore-PersonalizePreCreateScheduledRule'
-      Description: Calls Personalize pre-create Lambda function every 5 minutes until campaigns are fully created
+      Description: Calls Personalize pre-create Lambda function every 5 minutes until Personalize reaches desired state
       ScheduleExpression: rate(5 minutes)
-      State: ENABLED
+      State: DISABLED
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:214]: The creation of Personalize resources is initiated from CloudFormation as it calls the Custom Resource Lambda with appropriate resources. The event rule is then enabled by the Custom Resource Lambda so we do not mind putting this back in, we just feel enabling it is best coordinated by the Custom Resource itself.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 15]
```diff
@@ -159,48 +162,33 @@ Resources:
...
                 Action:
+                  - lambda:AddPermission
+                  - lambda:RemovePermission
+                  - events:PutRule
+                  - events:PutTargets
                   - events:ListTargetsByRule
                   - events:RemoveTargets
                   - events:DeleteRule
-                Resource:
-                  - !Sub 'arn:aws:events:${AWS::Region}:${AWS::AccountId}:rule/RetailDemoStore-PersonalizePreCreateScheduledRule'
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:203]: Will do

**Comment by daemon-joe** [comment]
> This is a massive PR and probably should have been done in phases. But we'll get through it. Excited to see IVS integrated! Some great work here.
> 
> General questions:
> 
> * Why has the delete personalize resources CR been deleted? How will Personalize resources be cleaned up when the user deletes the CFN stack?
> * Should the 1.3 personalized discounts notebook be 1.2 and the cleanup notebook renamed to 1.3? The idea of the cleanup notebook is to cleanup after you've finished the other notebook(s) in the folder.
> * The discount notebook seems to be more focused on validating the interaction generation methodology than as an educational tool for customers. For example, all of the steps a customer would normally want to learn on how to use the Personalize service is hidden away in the lambda function. I'm also wondering if the 4th campaign is necessary. It's not completely clear what its purpose is. Could contextual metadata have been used instead with the existing campaigns?
> 
> See more detailed questions throughout PR.
> 
> Thanks!

Thanks for the feedback James and for taking the time to discuss with us yesterday. I've replied to the code-specific comments above where appropriate and I'll try to give answers to the more general questions here. I'm sure these will be the basis for some of our discussions going forward as well

1. In order to implement the "Personalize reset" button we needed to both delete and create Personalize resources so we shifted the delete functionality to the one pre-create-personalize-campaigns.py Lambda (re-using existing code where possible). We hoped that by combining the management of the Personalize infrastructure into a single Custom Resource we could simplify the management of the Personalize resources as well. 

2. Because this notebook does not currently build resources, the order does not matter a lot. If the notebook is expanded to build the discount campaigns, then the order will matter.

3a. We can certainly expand out the notebook and put the interactions generation and Personalize building directly into it. For the latter (building Personalize), the required steps are already tackled in notebook 1.1 so we opted to use the deployed pre-create lambda from this notebook to show how that might be used. For the former (generating interactions) we felt like explaining the additional logic for simulating interactions may be counterproductive from a perspective of teaching Personalize use. 

3b. We agree that, though we have not validated it, it could have been used and would have decreased the number of campaigns as suggested and shown off this feature of Personalize as a part of a series of Personalize workshops.
We note that there are a lot of different ways of handling discounts within Personalize and the decision we made was based on a desire to keep the IVS demo standalone alongside with Personalized Discounts.

**Comment by daemon-joe** [comment]
> Additional questions:
> 
> * The IVS integration looks to only support pre-recorded videos. How can a live stream be done using the integration?
> * Will there be a workshop notebook added to walk a developer through how to add IVS to the Retail Demo Store and run a live stream?
> * Given the scope of changes, have you done regression testing against the full application?
> * The delivery report (separate doc) indicates that CFN deploys sometimes fail and that it is a pre-existing issue. Please provide details on failures you have experienced so they can be fixed. There are no known/tracked issues with deployment at this time.
> 
> Thanks!

1. The current integration doesn’t support live streaming, since the Videos service currently returns only the details of the channels created by CloudFormation to stream the recorded video. 
To support live streaming we would have to make some tweaks to the videos service to allow a configurable extra channel to be used for live streaming by a user. In this case, it would also be best to make some UI changes to handle the case when there is no product metadata being sent alongside the stream (unless this functionality was also built into some sort of streaming interface). 

2. One for discussion on a call with yourself, Krithika and our team. 

3. We could not find any regression tests in the code base but we tested a variety of functionality including:

- Product display and search.
- Recommendations - user, item, search reranking.
- Default recommendations.
- Profile save + authentication.
- Checkout.
- Experimentation in recommendations.

We are currently testing:

- Pinpoint events/messaging (but we do not expect any surprises there).

4. To be clear, it is not necessarily a problem with the CloudFormation templates, but there was a problem whereby the ElasticSearch index failed to deploy due to a service issue in the regions we were using (we had a few issues with sticky resources, long creation times etc.). We do not have clear steps to reproduce, since after a few days ElasticSearch appeared to be functioning properly again.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 16]
```diff
@@ -104,7 +107,6 @@ Resources:
 
   PersonalizePreCreateLambdaFunction:
     Condition: DeployPreCreateCampaign
-    DependsOn: PersonalizeDeleteLambdaFunction # Ensure PreCreate function is deleted before Delete function so PreCreate is stopped first
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:107]: Have answered in the response to general questions [here](https://github.com/aws-samples/retail-demo-store/pull/96#issuecomment-714587413).

[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:107]: We have tested the teardown during creation and it deletes successfully - the delete function will keep running until the desired state (deleted) is reached.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 17]
```diff
@@ -243,27 +231,37 @@ Resources:
...
+  # Custom resource to launch IVS create channels function
+  CustomLaunchPersonalizePreCreateLambdaFunction:
+    Condition: DeployPreCreateCampaign
+    Type: Custom::CustomLambdaPersonalize
+    Properties:
+      ServiceToken: !GetAtt PersonalizePreCreateLambdaFunction.Arn
+
+  ####################### Create IVS Channels #######################
 
-  PersonalizeDeleteLambdaFunction:
```
[the remark this answers is not in the document]
Comment (daemon-joe) [reply aws/cloudformation-templates/deployment-support.yaml:248]: Have answered in the response to general questions [here](https://github.com/aws-samples/retail-demo-store/pull/96#issuecomment-714587413).

File: src/recommendations/src/recommendations-service/experimentation/resolvers.py [hunk 18]
```diff
@@ -351,6 +372,82 @@ def get_items(self, **kwargs):
 
         return echo_items
 
+
+class PersonalizeContextComparePickResolver(Resolver):
+    """ Provides personalized ranking of products from an Amazon Personalize campaign
+
+    The campaign must be trained using the Personalized-Ranking recipe
+    """
+    __personalize_runtime = boto3.client('personalize-runtime')
```
Comment (james-jory) [inline src/recommendations/src/recommendations-service/experimentation/resolvers.py:381]: Does not appear to be used.

File: src/recommendations/src/recommendations-service/experimentation/resolvers.py [hunk 19]
```diff
@@ -351,6 +372,82 @@ def get_items(self, **kwargs):
...
+
+        discount_improve_sorted_items = [with_id_to_item[item_id] for item_id in discount_improve_sorted_item_ids]
+
+        return discount_improve_sorted_items[:top_n]
+
+
+class RandomPickResolver(Resolver):
+    """ Picks random N products.
+    """
+    __personalize_runtime = boto3.client('personalize-runtime')
```
Comment (james-jory) [inline src/recommendations/src/recommendations-service/experimentation/resolvers.py:420]: Does not appear to be used.

File: src/videos/src/videos-service/app.py (hunk 2 above)
Comment (james-jory) [reply src/videos/src/videos-service/app.py:37]: Does it make sense to externalize the default stream details in a JSON file or something? Otherwise we're going to have to make changes to source when the product IDs change or when new/different default streams are added. Product IDs will be changing to UUIDs soon. I get that the defaults shouldn't change often but having the hardcoded makes it harder to maintain.

**Review by james-jory — CHANGES_REQUESTED** [review]
Looking pretty good. Just some specific comments/changes.

Also, be sure to resize the screenshot added to the Personalize workshop so it matches the resolution of the other screenshots. This will keep the repo size down (and below the 20MB max for seeding the CodeCommit repo).

File: aws/cloudformation-templates/services/service/_template.yaml [hunk 20]
```diff
@@ -178,6 +195,10 @@ Parameters:
     Type: String
     Description: Lambda Arn for deleting ECR repository
 
+  UseDefaultIvsStreams:
```
Comment (alexchirayath) [inline aws/cloudformation-templates/services/service/_template.yaml:198]: NIT: Can we use a standard convention across with either uppercase IVS everywhere or go with Ivs?

File: aws/cloudformation-templates/base/notebook.yaml [hunk 21]
```diff
@@ -118,7 +118,7 @@ Resources:
                 Action: 
                   - ssm:PutParameter
                   - ssm:GetParameter
-                Resource: !Sub 'arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/retaildemostore-*'
+                Resource: !Sub 'arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/retaildemo*'
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply aws/cloudformation-templates/base/notebook.yaml:121]: Yes you are right. This is unneeded. It is a change needed for the next demo as we have different dataset groups. We will remove it from this iteration.

File: aws/cloudformation-templates/base/ssm.yaml [hunk 22]
```diff
@@ -44,6 +44,14 @@ Resources:
       Value: !Ref StackBucketName
       Description: "Retail Demo Store Stack Bucket Parameter"
 
+  ParameterPersonalizeDatasetGroupName:
+    Type: "AWS::SSM::Parameter"
+    Properties:
+      Name: "retaildemostore-personalize-datasetgroup-name"
+      Type: "String"
+      Value: "NONE"
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply aws/cloudformation-templates/base/ssm.yaml:52]: We will remove this entirely as it is no longer used.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 23]
```diff
@@ -123,6 +125,7 @@ Resources:
           csv_bucket: !Ref ResourceBucket
           csv_path: !Sub '${ResourceBucketRelativePath}csvs/'
           lambda_event_rule_name: 'RetailDemoStore-PersonalizePreCreateScheduledRule'
+          num_to_train: '1'
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply aws/cloudformation-templates/deployment-support.yaml:128]: Again, you are absolutely right - this is left over from earlier versions of our solution - we needed to adapt the approach over time.

File: src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py [hunk 24]
```diff
@@ -29,12 +29,15 @@
...
+import uuid
+from crhelper import CfnResource
 
 from packaging import version
 from botocore.exceptions import ClientError
 
+
 logger = logging.getLogger()
-logger.setLevel(logging.INFO)
+logger.setLevel(logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py:40]: Will correct.

File: src/recommendations/src/recommendations-service/app.py [hunk 25]
```diff
@@ -305,122 +323,343 @@ def recommendations():
...
-            return resp
-    
-        except Exception as e:
-            app.logger.exception('Unexpected error reranking items', e)
-            return json.dumps(items) 
+@app.route('/choose_discounted', methods=['POST'])
+def choose_discounted():
+    """
+    Gets user ID, items list and feature and chooses which items to discount according to the discount reranking
+    campaign. Gets a ranking with discount applied and without and looks at the difference. The products
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply src/recommendations/src/recommendations-service/app.py:623]: Will correct.

File: src/recommendations/src/recommendations-service/app.py [hunk 26]
```diff
@@ -452,16 +691,118 @@ def experiment_outcome():
...
+def get_event_tracker_arn_from_id(dataset_group_arn, event_tracking_id):
+    """Looks for the Arn of an event tracker given its dataset group and Id"""
+    list_response = personalize.list_event_trackers(datasetGroupArn=dataset_group_arn, maxResults=100)
+    for event_tracker in list_response['eventTrackers']:
+        arn = event_tracker['eventTrackerArn']
+        describe_response = personalize.describe_event_tracker(eventTrackerArn=arn)
+        tracking_id = describe_response['eventTracker']['trackingId']
+        if tracking_id == event_tracking_id:
+            return arn
+    return None
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply src/recommendations/src/recommendations-service/app.py:723]: Will remove these functions. Again, we needed to adapt our approach several times.

File: src/recommendations/src/recommendations-service/experimentation/resolvers.py [hunk 27]
```diff
@@ -205,6 +209,9 @@ def get_items(self, **kwargs):
             elif self.filter_arn:
                 params['filterArn'] = self.filter_arn
 
+            # contextual metadata does not work in related items recipe
```
[the remark this answers is not in the document]
Comment (damiendaemon) [reply src/recommendations/src/recommendations-service/experimentation/resolvers.py:212]: Will do.

File: src/recommendations/src/recommendations-service/experimentation/resolvers.py (hunk 18 above)
Comment (damiendaemon) [reply src/recommendations/src/recommendations-service/experimentation/resolvers.py:381]: Will remove.

File: src/recommendations/src/recommendations-service/experimentation/resolvers.py (hunk 19 above)
Comment (damiendaemon) [reply src/recommendations/src/recommendations-service/experimentation/resolvers.py:420]: We will make this change. Default streams will be picked up from the `videos/` folder where user-provided videos may be provided.

File: aws/cloudformation-templates/services/service/_template.yaml (hunk 20 above)
Comment (damiendaemon) [reply aws/cloudformation-templates/services/service/_template.yaml:198]: We will update the CloudFormation to use uppercase IVS.

File: src/recommendations/src/recommendations-service/experimentation/resolvers.py (hunk 19 above)
Comment (damiendaemon) [reply src/recommendations/src/recommendations-service/experimentation/resolvers.py:420]: Will remove.

File: src/videos/src/videos-service/app.py (hunk 2 above)
Comment (damiendaemon) [reply src/videos/src/videos-service/app.py:37]: We will make this change. Default streams will be picked up from the videos/ folder where user-provided videos may be provided.

**Comment by damiendaemon** [comment]
Hi all. We have made the latest requests changes and updated this PR. We also needed to bump the version of Golang in the products service (because of: https://github.com/jinzhu/copier/issues/68 ) to allow this to deploy. Take care.

**Comment by james-jory** [comment]
> Hi all. We have made the latest requests changes and updated this PR. We also needed to bump the version of Golang in the products service (because of: [jinzhu/copier#68](https://github.com/jinzhu/copier/issues/68) ) to allow this to deploy. Take care.

Changes look good. Just need to address conflict with the copier fix in the products service. We have an issue to move all golang services to use modules so thanks for doing this for the products service. A PR was merged yesterday to remove copier as a dependency of this service as a quick fix. Either back out your changes to the products service or merge your changes with upstream (to remove copier and add your module improvement).

Thanks!

**Review by james-jory — APPROVED** [review]
LGTM. Thanks!
